core/context.cpp

```diff
@@ -87,7 +87,7 @@ void ContextBase::allocVoiceChanges()
         cluster[i-1].mNext.store(std::addressof(cluster[i]), std::memory_order_relaxed);
     cluster[clustersize-1].mNext.store(mVoiceChangeTail, std::memory_order_relaxed);
 
-    mVoiceChangeClusters.emplace_back(std::move(cluster));
+    mVoiceChangeClusters.push_back(std::move(cluster));
```

Comment:
Is there a reason for this change? `push_back` causes a new object to be default-constructed first and then the given parameter is assigned to it, which is less efficient than `emplace_back` which constructs the new object with the given parameter.

Reply:
push_back with std::move is the same really.

Reply:
In this particular case after the optimizer takes a stab at it, but not conceptually or in the general case. `push_back` first allocates space for and default-constructs a new object before anything else, then assigns the given parameter to the constructed object (default construction + move assignment). `emplace_back` allocates space for the object, and then constructs the new object using the parameter (move construction). As objects get more complicated, any guarantee of the compiler optimizing `push_back` to act like `emplace_back` goes down, so it's generally better to use `emplace_back` rather than hoping the compiler manages to optimize `push_back` as good as `emplace_back`.

Reply:
removed commit.